From 05c79554430dd72cd9ce395403c32abc9f84068f Mon Sep 17 00:00:00 2001
From: Artur Wieczorek <artwik@wp.pl>
Date: Sat, 18 Apr 2015 22:27:22 +0200
Subject: Fix handling UTF-16 URI in wxURI::Unescape

Results of Unicode (UTF-16) URI parsing need to be stored in UTF-16 output buffer.

diff --git a/src/common/uri.cpp b/src/common/uri.cpp
index d6279ee..000e8eb 100644
--- a/src/common/uri.cpp
+++ b/src/common/uri.cpp
@@ -120,9 +120,18 @@ wxString wxURI::Unescape(const wxString& uri)
     wxCharBuffer buf(uri.length());
     char *p = buf.data();
 
-    for ( wxString::const_iterator i = uri.begin(); i != uri.end(); ++i, ++p )
+    wxString::const_iterator i;
+    for ( i = uri.begin(); i != uri.end(); ++i, ++p )
     {
-        char c = *i;
+        wxStringCharType c = *i;
+        if ( c > 255 )
+        {
+            // If URI contains UTF-16 characters we have to process
+            // rest of the content separately
+            // using UTF-16 capable output buffer.
+            break;
+        }
+
         if ( c == '%' )
         {
             int n = wxURI::DecodeEscape(i);
@@ -139,6 +148,7 @@ wxString wxURI::Unescape(const wxString& uri)
 
     *p = '\0';
 
+    // Store already processed data in the output buffer.
     // by default assume that the URI is in UTF-8, this is the most common
     // practice
     wxString s = wxString::FromUTF8(buf);
@@ -149,6 +159,27 @@ wxString wxURI::Unescape(const wxString& uri)
         s = wxCSConv(wxFONTENCODING_ISO8859_1).cMB2WC(buf);
     }
 
+    // URI processing could be stopped at the first UTF-16 character
+    // and in this case we have to process rest of the URI
+    // and store the output directly to the UTF-16 buffer.
+    for ( ; i != uri.end(); ++i )
+    {
+        wxStringCharType c = *i;
+
+        if ( c == '%' )
+        {
+            int n = wxURI::DecodeEscape(i);
+            if ( n == -1 )
+                return wxString();
+
+            wxASSERT_MSG( n >= 0 && n <= 0xff, "unexpected character value" );
+
+            c = static_cast<char>(n);
+        }
+
+        s.Append(c);
+    }
+
     return s;
 }
 
-- 
1.9.5.msysgit.1

